src/chunkformatter.js

```diff
@@ -0,0 +1,311 @@
+ * CELL_IN_PROGRESS: state when valueSize > 0(partial cell)
+ */
+const RowStateEnum = Object.freeze({
+  NEW_ROW: 1,
+  ROW_IN_PROGRESS: 2,
+  CELL_IN_PROGRESS: 3,
+});
+/**
+ * ChunkFormatter formats all incoming chunks in to row
+ * keeps all intermediate state until end of stream.
+ * Should use new instance for each request.
+ */
+function ChunkFormatter() {
+  if (!(this instanceof ChunkFormatter)) {
+    return new ChunkFormatter();
+  }
+  this.reset();
+}
+/**
+ * Formats the row chunks into friendly format. Chunks contain 3 properties:
+ *
+ * `rowContents` The row contents, this essentially is all data pertaining
+ *     to a single family.
+ *
+ * `commitRow` This is a boolean telling us the all previous chunks for this
+ *     row are ok to consume.
+ *
+ * `resetRow` This is a boolean telling us that all the previous chunks are to
+ *     be discarded.
+ *
+ * @public
+ *
+ * @param {chunk[]} chunks The list of chunks.
+ * @param {object} [options] Formatting options.
+ * @param {callback} callback callback will be called with parsed rows
+ * @throws Error if it detects invalid state
+ * @example
+ * ChunkFormatter.formatChunks(chunks);
+ * // {
+ * //   follows: {
+ * //     gwashington: [
+ * //       {
+ * //         value: 2
+ * //       }
+ * //     ]
+ * //   }
+ * // }
+ */
+ChunkFormatter.prototype.formatChunks = function(chunks, options, callback) {
+  options = options || {};
+  for (const chunk of chunks) {
+    switch (this.state) {
+      case RowStateEnum.NEW_ROW:
+        this.newRow(chunk, options, callback);
+        break;
+      case RowStateEnum.ROW_IN_PROGRESS:
+        this.rowInProgress(chunk, options, callback);
+        break;
+      case RowStateEnum.CELL_IN_PROGRESS:
+        this.cellInProgress(chunk, options, callback);
+        break;
+    }
+  }
+};
+/**
+ * should be called at end of the stream to check if there is any pending row.
+ * @public
+ * @throws Error if there is any uncommitted row.
+ */
+ChunkFormatter.prototype.onStreamEnd = function() {
+  if (typeof this.row.key !== 'undefined') {
+    throw new Error('Response ended with pending row without commit');
+  }
+};
+/**
+ * Resets state of formatter
+ * @private
+ */
+ChunkFormatter.prototype.reset = function() {
+  this.prevRowKey = '';
+  this.family = {};
+  this.qualifiers = [];
+  this.qualifier = {};
+  this.row = {};
+  this.state = RowStateEnum.NEW_ROW;
+};
+/**
+ * sets prevRowkey and calls reset when row is committed.
+ * @private
+ */
+ChunkFormatter.prototype.commit = function() {
+  const row = this.row;
+  this.reset();
+  this.prevRowKey = row.key;
+};
+/**
+ * Error checker if true throws error
+ * @private
+ * @param {boolean} condition condition to evaluate
+ * @param {string} text Error text
+ * @param {object} chunk chunk object to append to text
+ */
+ChunkFormatter.prototype.isError = function(condition, text, chunk) {
+  if (condition) {
+    throw new Error(`${text}: ${chunk}`);
```

Comment:
`${chunk}` will just convert to `[object Object]` To have a meaningful information about chunk, maybe use `${JSON.stringify(chunk)}` or something else.

Reply:
updated